twoliter/src/cargo_make.rs

```diff
@@ -0,0 +1,213 @@
+use std::path::PathBuf;
+
+use anyhow::{bail, Result};
+use log::trace;
+use tokio::process::Command;
+
+use crate::common::exec;
+use crate::docker::ImageArchUri;
+use crate::project::Project;
+
+fn require_sdk(project: &Project, arch: &str) -> Result<(ImageArchUri, ImageArchUri)> {
+    match (project.sdk(arch), project.toolchain(arch)) {
+        (Some(s), Some(t)) => Ok((s, t)),
+        _ => bail!(
+            "When using twoliter make, it is required that the SDK and toolchain be specified in \
+            Twoliter.toml"
+        ),
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct CargoMake {
+    makefile_path: Option<PathBuf>,
+    project_dir: Option<PathBuf>,
+    args: Vec<String>,
+}
+
+impl CargoMake {
+    pub fn new<S>(project: &Project, arch: S) -> Result<Self>
+    where
+        S: Into<String>,
+    {
+        let (sdk, toolchain) = require_sdk(project, &arch.into())?;
+        Ok(Self::default()
+            .env("TLPRIVATE_SDK_IMAGE", sdk)
+            .env("TLPRIVATE_TOOLCHAIN", toolchain))
+    }
+
+    pub fn makefile<P>(mut self, makefile_path: P) -> Self
+    where
+        P: Into<PathBuf>,
+    {
+        self.makefile_path = Some(makefile_path.into());
+        self
+    }
+
+    pub fn project_dir<P>(mut self, project_dir: P) -> Self
+    where
+        P: Into<PathBuf>,
+    {
+        self.project_dir = Some(project_dir.into());
+        self
+    }
+
+    pub fn env<S1, S2>(mut self, key: S1, value: S2) -> Self
+    where
+        S1: Into<String>,
+        S2: Into<String>,
+    {
+        self.args
+            .push(format!("-e={}={}", key.into(), value.into()));
+        self
+    }
+
+    pub fn _envs<S1, S2, V>(mut self, env_vars: V) -> Self
+    where
+        S1: Into<String>,
+        S2: Into<String>,
+        V: Into<Vec<(S1, S2)>>,
+    {
+        for (key, value) in env_vars.into() {
+            self.args
+                .push(format!("-e={}={}", key.into(), value.into()));
+        }
+        self
+    }
+
+    pub fn _arg<S>(mut self, arg: S) -> Self
+    where
+        S: Into<String>,
+    {
+        self.args.push(arg.into());
+        self
+    }
+
+    pub fn _args<V, S>(mut self, args: V) -> Self
+    where
+        S: Into<String>,
+        V: Into<Vec<S>>,
+    {
+        self.args.extend(args.into().into_iter().map(Into::into));
+        self
+    }
+
+    pub async fn _exec<S>(&self, task: S) -> Result<()>
+    where
+        S: Into<String>,
+    {
+        self.exec_with_args(task, Vec::<String>::new()).await
+    }
+
+    pub async fn exec_with_args<S1, S2, V>(&self, task: S1, args: V) -> Result<()>
```

Comment:
Also something like `IntoIter<Item = String>` seems more common.

```suggestion
    /// Blah blah blah, explain what this does.
    pub(crate) async fn exec<S1, S2, V>(&self, task: S1, args: V) -> Result<()>
```